src/components/sticker.rs

```diff
@@ -0,0 +1,191 @@
+use adw::prelude::*;
+use adw::subclass::prelude::*;
+use glib::clone;
+use gtk::{gio, glib};
+
+use tdlib::enums::StickerFormat;
+use tdlib::types::Sticker as TdSticker;
+
+use crate::session::Session;
+use crate::utils::{decode_image_from_path, spawn};
+
+mod imp {
+    use super::*;
+    use std::cell::{Cell, RefCell};
+
+    #[derive(Default)]
+    pub struct Sticker {
+        pub(super) message_id: Cell<i64>,
+        pub(super) aspect_ratio: Cell<f64>,
+        pub(super) max_size: Cell<i32>,
+
+        pub(super) child: RefCell<Option<gtk::Widget>>,
+    }
+
+    #[glib::object_subclass]
+    impl ObjectSubclass for Sticker {
+        const NAME: &'static str = "ComponentsSticker";
+        type Type = super::Sticker;
+        type ParentType = gtk::Widget;
+    }
+
+    impl ObjectImpl for Sticker {}
+
+    impl WidgetImpl for Sticker {
+        fn measure(&self, orientation: gtk::Orientation, _for_size: i32) -> (i32, i32, i32, i32) {
+            let size = self.max_size.get();
+            let aspect_ratio = self.aspect_ratio.get();
+
+            let min_size = 1;
+
+            let size = if let gtk::Orientation::Horizontal = orientation {
+                if aspect_ratio >= 1.0 {
+                    size
+                } else {
+                    (size as f64 * aspect_ratio) as i32
+                }
+            } else if aspect_ratio >= 1.0 {
+                (size as f64 / aspect_ratio) as i32
+            } else {
+                size
+            }
+            .max(min_size);
+
+            (size, size, -1, -1)
+        }
+
+        fn size_allocate(&self, width: i32, height: i32, baseline: i32) {
+            if let Some(child) = &*self.child.borrow() {
+                child.allocate(width, height, baseline, None);
+            }
+        }
+
+        fn snapshot(&self, snapshot: &gtk::Snapshot) {
+            if let Some(child) = &*self.child.borrow() {
+                self.obj().snapshot_child(child, snapshot);
+            }
+        }
+
+        fn request_mode(&self) -> gtk::SizeRequestMode {
+            gtk::SizeRequestMode::HeightForWidth
+        }
```

Comment:
I'm fairly sure you don't need `HeightForWidth` here. I think you need a constant size, so you can probably just delete this function.